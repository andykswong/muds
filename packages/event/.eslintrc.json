{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "tsconfig.json",
    "sourceType": "module"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended"
  ],
  "env": {
    "es2022": true,
    "browser": true,
    "node": true,
    "jest": true
  },
  "ignorePatterns": [
  ],
  "rules": {
  }
}
