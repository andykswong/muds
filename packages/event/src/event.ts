/** Listener of events. */
export interface EventListener<T extends unknown[]> {
  (...data: T): void;
}

/** A typesafe event signal. */
export interface Event<T extends unknown[]> {
  /** Adds an event listener. */
  addListener(listener: EventListener<T>): void;

  /** Removes an event listener. */
  removeListener(listener: EventListener<T>): void;

  /** Emits an event. */
  emit(...data: T): void;
}

/** A typesafe event signal. */
export class Event<T> implements Event<T> {
  /** Creates a simple event instance. */
  public static create<T extends unknown[]>(): Event<T> {
    return new SimpleEvent();
  }
}

/** A simple Event implementation that triggers listeners immediately upon emitting. */
export class SimpleEvent<T extends unknown[]> implements Event<T> {
  private readonly listeners: EventListener<T>[] = [];

  public addListener(listener: EventListener<T>) {
    this.listeners.push(listener);
  }

  public removeListener(listener: EventListener<T>) {
    const index = this.listeners.indexOf(listener);
    if (index > -1) {
      this.listeners[index] = this.listeners[this.listeners.length - 1];
      this.listeners.pop();
    }
  }

  public emit(...data: T) {
    for (const listener of this.listeners) {
      listener(...data);
    }
  }
}
