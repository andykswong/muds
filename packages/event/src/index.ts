export * from './event';
export * from './multiplexer';
export * from './time';
