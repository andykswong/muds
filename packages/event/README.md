# @muds/event
<p align="center">
  <a href="https://www.npmjs.com/package/@muds/event"><img src="https://img.shields.io/npm/v/@muds/event?label=@muds/event" alt="NPM @muds/event" /></a>
  <a href="./LICENSE"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT" /></a>
</p>

## Overview
`@muds/event` is the core event signal library for [muds](https://github.com/andykswong/muds), the modular microframework for interactive data-oriented systems. It provides the event emitter and listener interface, as well as several useful event implementations.

## API Documentation
http://andykswong.github.io/muds

## Usage
```shell
npm install --save @muds/event
```

```javascript
import { Event, EventMultiplexer } from '@muds/event';

const listener = (...msg) => console.log(...msg);
const event1 = Event.create(); // simple event
const event2 = Event.create();

event1.addListener(listener);
event1.emit('hello', 'world'); // log "hello world"
event1.removeListener(listener);

const multiplexer = new EventMultiplexer(); // multiplex events into a single event
multiplexer.addListener(listener);
multiplexer.add(event1, (event) => [1, ...event]); // add an event with event mapping
multiplexer.add(event2, (event) => [2, ...event]);
event1.emit('hello', 'again'); // log "1 hello again"
event2.emit('bye'); // log "2 bye"
```

```javascript
import { OnAnimationFrame } from '@muds/event';

const listener = (delta) => console.log(`delta: ${delta} ms`);
const event = new OnAnimationFrame(); // emit event with delta time on each requestAnimationFrame callback
event.addListener(listener);
event.start(); // log "delta: 16ms" periodically
// event.pause(); to stop
```

## License
This repository and the code inside it is licensed under the MIT License. Read [LICENSE](./LICENSE) for more information.
