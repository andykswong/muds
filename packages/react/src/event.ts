import { useState, useEffect, useRef } from 'react';
import { Event, EventListener } from '@muds/event';

/** Hook to listen to an event. */
export function useEventListener<E extends unknown[]>(
  event: Event<E>,
  listener: EventListener<E>,
): void {
  const savedListener = useRef<EventListener<E>>();

  useEffect(() => {
    savedListener.current = listener;
  }, [listener]);

  useEffect(() => {
    function onEvent(...event: E) {
      savedListener.current?.(...event);
    }
    event.addListener(onEvent);
    return () => event.removeListener(onEvent);
  }, [event]);
}

/** Hook to derive state from event. */
export function useEventReducer<S, E extends unknown[]>(
  event: Event<E>,
  reducer: (state: S, ...event: E) => S,
  initialState: S | (() => S),
): S {
  const [state, setState] = useState(initialState);
  const onEvent = useRef<EventListener<E>>();

  useEffect(() => {
    onEvent.current = (...event: E) => {
      setState(reducer(state, ...event));
    };
  }, [state, setState, reducer]);

  useEffect(() => {
    const listener = (...event: E) => onEvent.current?.(...event);
    event.addListener(listener);
    return () => event.removeListener(listener);
  }, [event]);

  return state;
}
