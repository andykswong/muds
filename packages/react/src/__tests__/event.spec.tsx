import { jest } from '@jest/globals';
import React, { useCallback, useState } from 'react';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import { Event } from '@muds/event';
import { useEventListener, useEventReducer } from '../event';

enum Action {
  Increment = 'inc',
  Decrement = 'dec',
}

function reducer(count: number, action: Action): number {
  switch (action) {
    case Action.Increment: return count + 1;
    case Action.Decrement: return count - 1;
    default: return count;
  }
}

describe('useEventListener', () => {
  function Counter({ event, initialCount = 0 }: { event: Event<[Action]>, initialCount?: number }) {
    const [count, setCount] = useState(initialCount);
    useEventListener(event, useCallback((action) => setCount(reducer(count, action)), [count, setCount]));
    return (<span>{count}</span>);
  }

  it('should add/remove event listener on mount/unmount', () => {
    const event = Event.create<[Action]>();

    const addListener = jest.spyOn(event, 'addListener');
    const removeListener = jest.spyOn(event, 'removeListener');

    let component!: ReactTestRenderer;

    act(() => {
      component = create(<Counter event={event} />);
    });

    expect(addListener).toHaveBeenCalled();
    const listener = addListener.mock.calls[0][0];

    act(() => component.unmount());
    expect(removeListener).toHaveBeenCalledWith(listener);
  });

  it('should add listener to event', () => {
    const event = Event.create<[Action]>();

    let component!: ReactTestRenderer;

    act(() => {
      component = create(<Counter event={event} />);
    });
    expect(component.toJSON()).toMatchSnapshot();

    for (const action of [Action.Increment, Action.Increment, Action.Decrement]) {
      act(() => event.emit(action));
      expect(component.toJSON()).toMatchSnapshot();
    }
  });
});

describe('useEventReducer', () => {
  function Counter({ event, initialCount = 0 }: { event: Event<[Action]>, initialCount?: number }) {
    const count = useEventReducer(event, reducer, initialCount);
    return (
      <span>
        <button onClick={() => event.emit(Action.Decrement)}>-</button>
        {` ${count} `}
        <button onClick={() => event.emit(Action.Increment)}>+</button>
      </span>
    );
  }

  it('should add/remove event listener on mount/unmount', () => {
    const event = Event.create<[Action]>();

    const addListener = jest.spyOn(event, 'addListener');
    const removeListener = jest.spyOn(event, 'removeListener');

    let component!: ReactTestRenderer;

    act(() => {
      component = create(<Counter event={event} />);
    });

    expect(addListener).toHaveBeenCalled();
    const listener = addListener.mock.calls[0][0];

    act(() => component.unmount());
    expect(removeListener).toHaveBeenCalledWith(listener);
  });

  it('should reduce events to state', () => {
    const event = Event.create<[Action]>();

    let component!: ReactTestRenderer;

    act(() => {
      component = create(<Counter event={event} />);
    });
    expect(component.toJSON()).toMatchSnapshot();

    for (let i = 0; i < 2; ++i) {
      act(() => {
        const [, inc] = component.root.findAllByType('button');
        inc.props.onClick();
      });
      expect(component.toJSON()).toMatchSnapshot();
    }
  });
});
