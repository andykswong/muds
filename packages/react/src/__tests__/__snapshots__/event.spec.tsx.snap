// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useEventListener should add listener to event 1`] = `
<span>
  0
</span>
`;

exports[`useEventListener should add listener to event 2`] = `
<span>
  1
</span>
`;

exports[`useEventListener should add listener to event 3`] = `
<span>
  2
</span>
`;

exports[`useEventListener should add listener to event 4`] = `
<span>
  1
</span>
`;

exports[`useEventReducer should reduce events to state 1`] = `
<span>
  <button
    onClick={[Function]}
  >
    -
  </button>
   0 
  <button
    onClick={[Function]}
  >
    +
  </button>
</span>
`;

exports[`useEventReducer should reduce events to state 2`] = `
<span>
  <button
    onClick={[Function]}
  >
    -
  </button>
   1 
  <button
    onClick={[Function]}
  >
    +
  </button>
</span>
`;

exports[`useEventReducer should reduce events to state 3`] = `
<span>
  <button
    onClick={[Function]}
  >
    -
  </button>
   2 
  <button
    onClick={[Function]}
  >
    +
  </button>
</span>
`;
